### ISSUE TITLE
Incorrect averaging in SparseCriterion

### ISSUE THREAD

**Issue by jumutc** [fstrub95/nnsparse#2]
Hello,

I have noticed that `nnsparse.SparseCriterion` contains incorrect averaging with respect to provided batches. The running count is across all non-zero elements rather than across batch size, i.e. `pairs(target)`. I have consistent generalization issues (for instance when using `nn.MSECriterion`) due to this averaging. 

If I replace `noElem = noElem + t:size(1)` with `noElem = noElem + 1` in `nnsparse.SparseCriterion` problem is solved!

Thanks in advance!

**Comment by fstrub95** [comment]
Hello, 
Actually, the average criterion does not aim at averaging over the batch size, but over the number of elements.

> MSECriterion
> `criterion = nn.MSECriterion()`
> Creates a criterion that measures the mean squared error between n elements in the input x and output y:
> `loss(x, y) = 1/n \sum |x_i - y_i|^2 .`
> If x and y are d-dimensional Tensors with a total of n elements, the sum operation still operates over all the elements, and divides by n. 
> source : https://github.com/torch/nn/blob/master/doc/criterion.md#nn.MSECriterion

Therefore, computing the average among non-zero elements make sense in the sparse case.

If you want to average over the batch size, I would advise you to use this kind of code. 
(Note the bacthSize in the return statement)

```
lossFct = nn.SparseCriterion(nn.MSECriterion())
lossFct.sizeAverage = False


      -- Define the closure to evalute w and dw
      local function feval(x)

         -- Reset gradients and losses
         network:zeroGradParameters()

         -- AutoEncoder targets
         local target = input

         --- FORWARD
         local output = network:forward(input)
         local loss   = lossFct:forward(output, target)

         --- BACKWARD
         local dloss = lossFct:backward(output, target)
         local _     = network:backward(noisyInput, dloss)

         -- Return loss and gradients
         return loss/batchSize, dw:div(batchSize)
      end

```

 For further details, you can have a look to the test : https://github.com/fstrub95/nnsparse/blob/master/test/testSparseCriterion.lua

**Comment by jumutc** [comment]
OK. Maybe I pointed to the wrong direction. In my network I'm using `nn.Linear` module as the last layer and in this case when the module takes a batch input I have some problems. Propagated from `nn.SparseCriterion` gradients are being summed up in `nn.Linear` w.r.t. the batch (usually the first) tensor dimension. And here is the tricky point: because the denominator in `nnsparse.SparseCriterion` is the total number of non-zeros the wrongly scaled `gradOutput` in `nn.Linear` is inducing the wrong summation across the processed batch. Correct me if I'm wrong ;)

**Comment by fstrub95** [comment]
You are right, the summation is wrong if the flag sizeAverage is true. My point is that you should not activate this flag during the training. 
sizeAverage is only useful during the testing phase.

Regarding the issue, you should use this combo
`lossFct.sizeAverage = False`
and
`return loss/batchSize, dw:div(batchSize)`

**Comment by jumutc** [comment]
Shouldn't then `sizeAverage = false` be a default choice for `nnsparse.SparseCriterion` because I've spent quite a bit of time to realize that fact??? In fact generalization on a validation set was also poor and that was my intention to investigate this problem!

**Comment by jumutc** [comment]
I can imagine also a fix based on `self.train` field of any `nn.Module`!

**Comment by fstrub95** [comment]
To be honest, I agree with you. sizeAverage should be false by default. Yet, Torch set it to true by default... (cf above). I wanted to be as consistent as possible as the torch framework. I also put a big WARNING in the doc ;) I also cost me several days realize that! 

In the evaluation step, I also advise you to remove the size average and to divide it by the number of non-zero at the end. Indeed,  if you use minibatch with the flag sizeAverage you may face the following issue
loss1/n1 + loss2/n2 with n1 != n2   is not equal to (loss1+loss2)/(n1+n2)
with n1 and n2 the number of non-zero element

PS What is the fix you were thinking of?

**Comment by jumutc** [comment]
Actually I've just realized that the solution is pretty simple: in `SparseCriterionCorrect:__updateOutput` code might stay the same... but in `SparseCriterionCorrect:__updateGradInput` changing the increment of `nElem` to `nElem = nElem + 1` fixes this issue! In this case consistency is preserved with respect to the computation in previous layers. So at least for one batch the computation should be correct. 

And indeed setting `sizeAverage = false` should fix the issue with averaging on the evaluation step!

**Comment by jumutc** [comment]
But the problem of evaluating the correct running average still remains. For now I'm avoiding it (with `sizeAverage = true`) by a simple trick:  `(loss_1*n_1 + loss_2*n_2 + .. + loss_k*n_k) / total_n`.

**Comment by fstrub95** [comment]
Again, you can use the same trick:
That would avoid some floating point issues.

```
local rmseFct = nnsparse.SparseCriterion(nn.MSECriterion())
rmseFct.sizeAverage = false

local noRatings = 0

for k, oneInput in pairs(train) do

     -- Focus on the prediction aspect
     local oneTarget = test[k]

     -- Compute the number of ratings
     noRatings  = noRatings + oneTarget:size(1)

         -- Update the rmse
     local output = network:forward(input)
     rmse = rmse + rmseFct:forward(output, target)            

end

rmse /= noRatings
```

Note this is one way to do among many others. In my code, I compute beforehand the number of ratings.

Be very careful while modifying the source code. Sparsity is very sneaky :)

**Comment by jumutc** [comment]
OK, but there are 2 problems at least which we should address:
1.  a bug related to incorrect averaging of backpropogated gradients (I can suggest a fix)
2.  a default averaging policy in `nnsparse.SparseCriterion` which could stay as it is right now but for the correct averaging across batches we could suggest something else, like an additional API to get summed losses/gradients or smth else

**Comment by jumutc** [comment]
Hi @fstrub95, I would like to fix this issue and propose a pull request... is it OK for you? I feel like we have stopped discussing this issue ;)

**Comment by fstrub95** [comment]
Hi, i am on holidays for now. May we delay it for a week please.
As I said, my main concern is to be consistent with the torch framework. 
Yet, I am always keen to add new features :)
If averaging the batch does exist in torch, I will update the source code with your pull
request. In the worst case, I may only cherry pick some of your change or provide a link to your fork.
Kind regards, 
STRUB Florian

**Comment by fstrub95** [comment]
I am back. Ready to enhance the framework :)

**Comment by jumutc** [comment]
OK, cool! I will submit my pull request shortly.

**Comment by jumutc** [comment]
Hi, I have a problem with running tests. Do you have a special code for running them? Otherwise without some slight adjustments (like defining proper paths to `src` folder and calling before `nnsparse = {}`) everything fails.

**Comment by fstrub95** [comment]
When I wrote the test, I had a few symlinks with the files. It would be
better to change the include path.
Yet, feel free to run the test as you want. I am fine if I cannot run them
out of the box

Le 21 sept. 2016 11:10 AM, "Vilen Jumutc" notifications@github.com a
écrit :

**Comment by jumutc** [comment]
Is it OK if I adjust the tests and the doc such that it is explained how-to-run the tests?

**Comment by jumutc** [comment]
Hi, I have also noticed that `sparseLinearBatch.miniBatch()` in `testSparseLinearBatch.lua` is failing with incorrect `gradWeight` being returned from `nnsparse.SparseLinearBatch`. I may create another issue w.r.t. this. Will create a pull request shortly.
